### ISSUE TITLE
Issues converting recurrent ML-Agents model to TFLite (Tensorflow 1.15)

### ISSUE THREAD

**Issue by dav1d-wright** [tensorflow/tensorflow#34035]
Hi

I have been tryingt to convert a frozen graph def created by [ML-Agents](https://github.com/Unity-Technologies/ml-agents) to TFLite. However, there is an input variable to the model, which is defined by:

```
self.sequence_length = tf.placeholder(
            shape=None, dtype=tf.int32, name="sequence_length"
        )
```

which seems to cause problems in the conversion.

If I convert without passing shapes the following happens:

```
tflite_convert --output_file=model.tflite --graph_def_file=unity/ml-agents/models/ppo-0/RoverLearning/frozen_graph_def.pb --input_arrays=vector_observation,epsilon,recurrent_in,sequence_length --output_arrays=action
```

```
2019-11-05 13:59:21.589574: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-11-05 13:59:21.602134: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x7fe2bf9e7280 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2019-11-05 13:59:21.602151: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
Traceback (most recent call last):
  File "/Users/dwright/anaconda3/envs/mlagents/bin/tflite_convert", line 8, in <module>
    sys.exit(main())
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/python/tflite_convert.py", line 515, in main
    app.run(main=run_main, argv=sys.argv[:1])
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/python/platform/app.py", line 40, in run
    _run(main=main, argv=argv, flags_parser=_parse_flags_tolerate_undef)
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/python/tflite_convert.py", line 511, in run_main
    _convert_tf1_model(tflite_flags)
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/python/tflite_convert.py", line 199, in _convert_tf1_model
    output_data = converter.convert()
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/python/lite.py", line 898, in convert
    self._set_batch_size(batch_size=1)
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/python/lite.py", line 1032, in _set_batch_size
    shape = tensor.shape.as_list()
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/python/framework/tensor_shape.py", line 1171, in as_list
    raise ValueError("as_list() is not defined on an unknown TensorShape.")
ValueError: as_list() is not defined on an unknown TensorShape.
```
I added some debut messages to `lite.py` and found that the issue is with the variable `sequence_length`

If I pass the shapes to the command like this: 

```
tflite_convert --output_file=model.tflite --graph_def_file=unity/ml-agents/models/ppo-0/RoverLearning/frozen_graph_def.pb --input_arrays=vector_observation,epsilon,recurrent_in,sequence_length --output_arrays=action --input_shapes=1,7:1,2:7,40:1
```

I get a similar error:

```
2019-11-05 14:25:44.658772: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2019-11-05 14:25:44.677569: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x7fee79ed3960 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2019-11-05 14:25:44.677591: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2019-11-05 14:25:44.741840: I tensorflow/core/grappler/devices.cc:60] Number of eligible GPUs (core count >= 8, compute capability >= 0.0): 0 (Note: TensorFlow was not compiled with CUDA support)
2019-11-05 14:25:44.741948: I tensorflow/core/grappler/clusters/single_machine.cc:356] Starting new session
2019-11-05 14:25:44.819310: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:786] Optimization results for grappler item: graph_to_optimize
2019-11-05 14:25:44.819337: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:788]   constant_folding: Graph size after: 304 nodes (-81), 371 edges (-83), time = 56.212ms.
2019-11-05 14:25:44.819342: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:788]   constant_folding: Graph size after: 304 nodes (0), 371 edges (0), time = 8.713ms.
Traceback (most recent call last):
  File "/Users/dwright/anaconda3/envs/mlagents/bin/tflite_convert", line 8, in <module>
    sys.exit(main())
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/python/tflite_convert.py", line 515, in main
    app.run(main=run_main, argv=sys.argv[:1])
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/python/platform/app.py", line 40, in run
    _run(main=main, argv=argv, flags_parser=_parse_flags_tolerate_undef)
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/absl/app.py", line 299, in run
    _run_main(main, args)
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/absl/app.py", line 250, in _run_main
    sys.exit(main(argv))
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/python/tflite_convert.py", line 511, in run_main
    _convert_tf1_model(tflite_flags)
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/python/tflite_convert.py", line 199, in _convert_tf1_model
    output_data = converter.convert()
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/python/lite.py", line 983, in convert
    **converter_kwargs)
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/python/convert.py", line 449, in toco_convert_impl
    enable_mlir_converter=enable_mlir_converter)
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/python/convert.py", line 200, in toco_convert_protos
    raise ConverterError("See console for info.\n%s\n%s\n" % (stdout, stderr))
tensorflow.lite.python.convert.ConverterError: See console for info.
2019-11-05 14:25:46.811578: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.811623: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.811639: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.811649: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.811675: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.811685: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.811695: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.811705: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.811986: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812009: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812022: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812033: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812378: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: TensorArrayV3
2019-11-05 14:25:46.812411: I tensorflow/lite/toco/import_tensorflow.cc:193] Unsupported data type in placeholder op: 20
2019-11-05 14:25:46.812433: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: TensorArrayV3
2019-11-05 14:25:46.812448: I tensorflow/lite/toco/import_tensorflow.cc:193] Unsupported data type in placeholder op: 20
2019-11-05 14:25:46.812468: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812497: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: TensorArrayV3
2019-11-05 14:25:46.812507: I tensorflow/lite/toco/import_tensorflow.cc:193] Unsupported data type in placeholder op: 20
2019-11-05 14:25:46.812525: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812544: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812559: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812567: I tensorflow/lite/toco/import_tensorflow.cc:193] Unsupported data type in placeholder op: 20
2019-11-05 14:25:46.812580: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812588: I tensorflow/lite/toco/import_tensorflow.cc:193] Unsupported data type in placeholder op: 20
2019-11-05 14:25:46.812613: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: TensorArrayScatterV3
2019-11-05 14:25:46.812631: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812640: I tensorflow/lite/toco/import_tensorflow.cc:193] Unsupported data type in placeholder op: 20
2019-11-05 14:25:46.812661: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: TensorArrayScatterV3
2019-11-05 14:25:46.812680: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812694: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812708: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812724: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Enter
2019-11-05 14:25:46.812751: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: LoopCond
2019-11-05 14:25:46.812763: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: LoopCond
2019-11-05 14:25:46.812864: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: Exit
2019-11-05 14:25:46.812928: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: TensorArrayReadV3
2019-11-05 14:25:46.812951: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: TensorArraySizeV3
2019-11-05 14:25:46.813037: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: TensorArrayReadV3
2019-11-05 14:25:46.813111: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: TensorArrayGatherV3
2019-11-05 14:25:46.813301: I tensorflow/lite/toco/import_tensorflow.cc:659] Converting unsupported operation: TensorArrayWriteV3
2019-11-05 14:25:46.816417: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 185 operators, 322 arrays (0 quantized)
2019-11-05 14:25:46.818204: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] After Removing unused ops pass 1: 103 operators, 180 arrays (0 quantized)
2019-11-05 14:25:46.819853: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 103 operators, 180 arrays (0 quantized)
2019-11-05 14:25:46.819987: F tensorflow/lite/toco/graph_transformations/propagate_fixed_sizes.cc:1616] Check failed: *packed_shape == shape All input arrays to Pack operators must have the same shape. Input "sequence_length" is different.
Fatal Python error: Aborted

Current thread 0x0000000111d5b5c0 (most recent call first):
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/toco/python/toco_from_protos.py", line 52 in execute
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/absl/app.py", line 250 in _run_main
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/absl/app.py", line 299 in run
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/python/platform/app.py", line 40 in run
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/toco/python/toco_from_protos.py", line 89 in main
  File "/Users/dwright/anaconda3/envs/mlagents/bin/toco_from_protos", line 8 in <module>
```

Any help is highly appreciated!

**Comment by kaploink** [comment]
Hi

I'm exploring the same rabbit hole (ml-agents -> tflite); currently at about the same point.

Here are a few learnings that may help:
- try the python api instead of command line util (it's recommended): https://www.tensorflow.org/lite/convert/python_api
- if you happen to be on windows, the windows distrib of tensorflow seems to be missing lite or just the converter, but seems to work fine running it in the linux sub system (e.g. ubuntu): https://docs.microsoft.com/en-us/windows/wsl/install-win10.
- perhaps try the older version of tf lite converter api, that matches the tf version used by ml-agents (1.7): https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/g3doc/r1/convert/python_api.md#additional-instructions
- maybe try reporting this in the ml-agents repo instead (as ml-agents is well aware of tensorflow, but likely less so vice versa)

If you get any further pls let me know - I'll do the same!

Cheers
Matt

**Comment by kaploink** [comment]
Ha, just found your post in ml-agents repo!

**Comment by kaploink** [comment]
Note https://www.tensorflow.org/lite/convert/rnn#not_currently_supported which mentions RNN usage of `sequence_length` is not supported, but will be by end of 2019...

Maybe it is supported in previous version of converter.

My case isn't an RNN btw, so sounds like your rabbit hole might be a bit longer.

**Comment by wrd90** [comment]
Wow, thanks @kaploink! I will look into those suggestions and keep you up to date! Could you tell me what your set up and objective is, just for information?

**Comment by mp-airwallex** [comment]
@wrd90 
I'm making a low cost quadruped robot running on arduino or raspberry pi (tbd), and using unity's ml-agents to teach it to walk.

I've got the physical prototype robot and the corresponding trained ml-agents model, now at the pointy end of trying to put them together...

Currently at the conversion to tflite stage.

How did you find out what the input arrays/shape should be...?

**Comment by wrd90** [comment]
That sounds cool @mp-airwallex! 

I found the shapes by visualizing `frozen_graph_def.pb` using tensorboard

**Comment by kaploink** [comment]
err, oops, wrong account!  kaploink == mp-airwallex

ah, awesome, thanks

**Comment by wrd90** [comment]
> Note https://www.tensorflow.org/lite/convert/rnn#not_currently_supported which mentions RNN usage of `sequence_length` is not supported, but will be by end of 2019...
> 
> Maybe it is supported in previous version of converter.
> 
> My case isn't an RNN btw, so sounds like your rabbit hole might be a bit longer.

I have tried to use the drop-in replacements suggested in your link but it looks like that doesn't work with the version used by ml-agents...

```
  File "/Users/dwright/anaconda3/envs/mlagents/bin/mlagents-learn", line 11, in <module>
    load_entry_point('mlagents', 'console_scripts', 'mlagents-learn')()
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/learn.py", line 417, in main
    run_training(0, run_seed, options, Queue())
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/learn.py", line 233, in run_training
    options.multi_gpu,
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/trainer_util.py", line 91, in initialize_trainers
    multi_gpu,
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/ppo/trainer.py", line 75, in __init__
    seed, brain, trainer_parameters, self.is_training, load
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/ppo/policy.py", line 47, in __init__
    brain, trainer_params, reward_signal_configs, is_training, load, seed
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/ppo/policy.py", line 97, in create_model
    trainer_params.get("vis_encode_type", "simple")
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/ppo/models.py", line 52, in __init__
    self.create_cc_actor_critic(h_size, num_layers, vis_encode_type)
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/ppo/models.py", line 91, in create_cc_actor_critic
    name="lstm_policy",
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/models.py", line 577, in create_recurrent_encoder
    rnn_cell, lstm_input_state, initial_state=lstm_vector_in
  File "/Users/dwright/anaconda3/envs/mlagents/lib/python3.6/site-packages/tensorflow_core/lite/experimental/examples/lstm/rnn.py", line 168, in dynamic_rnn
    raise RuntimeError("OpHint dynamic rnn only supports control flow v2.")
RuntimeError: OpHint dynamic rnn only supports control flow v2.
```

**Comment by wrd90** [comment]
@kaploink I almost figured ;) 

I tried to replace the `LSTMCell` with the drop-in replacements suggested in your link, but there seems to be a problem which sounds much like the initial error message in the initial conversion:

```
Sorting model, may take a while... Done!
IGNORED: Cast unknown layer
IGNORED: AddV2 unknown layer
IGNORED: Shape unknown layer
IGNORED: TensorListReserve unknown layer
IGNORED: TensorListFromTensor unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: EmptyTensorList unknown layer
IGNORED: While unknown layer
IGNORED: TensorListStack unknown layer
Traceback (most recent call last):
  File "/Users/dwright/anaconda3/envs/mlagents/bin/mlagents-learn", line 11, in <module>
    load_entry_point('mlagents', 'console_scripts', 'mlagents-learn')()
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/learn.py", line 417, in main
    run_training(0, run_seed, options, Queue())
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/learn.py", line 255, in run_training
    tc.start_learning(env)
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/trainer_controller.py", line 219, in start_learning
    self._export_graph()
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/trainer_controller.py", line 129, in _export_graph
    self.trainers[brain_name].export_model()
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/trainer.py", line 152, in export_model
    self.policy.export_model()
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/tf_policy.py", line 230, in export_model
    tf2bc.convert(frozen_graph_def_path, self.model_path + ".nn")
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/tensorflow_to_barracuda.py", line 1552, in convert
    i_model, args
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/tensorflow_to_barracuda.py", line 1377, in process_model
    process_layer(n, o_context, args)
  File "/Users/dwright/MSE/projects/rainSim/unity/ml-agents/ml-agents/mlagents/trainers/tensorflow_to_barracuda.py", line 1220, in process_layer
    assert all_elements_equal(input_ranks)
AssertionError
```

**Comment by kaploink** [comment]
@wrd90
Ah okay, well looks like the tf pros have been adding some labels so sounds like they might have some answers for you soon.

The tflite conversion worked fine for me (non-RNN) using a slightly adjusted version of your command (`tflite_convert --output_file=model.tflite --graph_def_file=./frozen_graph_def.pb --input_arrays=vector_observation,epsilon --output_arrays=action`), so thanks heaps for that.  I've been able to continue onto tflite inference test (worked :) ), and now chucking it on the pi and hooking up to sensors & motors.  Not bothering with microcontroller version yet - will optimise later.

Cheers

**Comment by wrd90** [comment]
@kaploink Glad to hear you've been able to get some inference results! I have decided to move away from LSTMs for now. Instead I am currently buffering my observations and actions, which I then feed into the network as well. I am hoping this will perform well enough, but so far I have not seen any significant differences in the learning process compared to the LSTM version. Due to time restrictions in my project I had to find a workaround for now. Also I expect the support for RNNs only to be added to TF2, to which ML-Agents will need to migrate first. 
Cheers
